### ISSUE TITLE
Having a problem with 5.4 GPG key

### ISSUE THREAD

**Issue by thomascooper** [docker-library/php#72]
During a standard build, using the Dockerfile in 5.4/apache I am getting the following error:
Step 12 : RUN gpg --keyserver pool.sks-keyservers.net --recv-keys F38252826ACD957EF380D39F2F7956BC5DA04B5D
 ---> Running in 2104b57cbaf4
gpg: directory `/root/.gnupg' created
gpg: new configuration file `/root/.gnupg/gpg.conf' created
gpg: WARNING: options in `/root/.gnupg/gpg.conf' are not yet active during this run
gpg: keyring `/root/.gnupg/secring.gpg' created
gpg: keyring `/root/.gnupg/pubring.gpg' created
gpg: requesting key 5DA04B5D from hkp server pool.sks-keyservers.net
?: pool.sks-keyservers.net: Host not found
gpgkeys: HTTP fetch error 7: couldn't connect: Success
gpg: no valid OpenPGP data found.
gpg: Total number processed: 0

I am spinning up the docker container on OSX using Vagrant 1.6.3 (ubuntu trusty64 image), if I use the image php:5.4-apache, i have no problems, but I want to modify the initial Dockerfile and build my own image. 

I may be totally mis-remembering but I believe this worked a few days ago. If I run the gpg line on another VM (ubuntu 14.04) I get the following:
tom@tom-VB14dev:~$ gpg --keyserver pool.sks-keyservers.net --recv-keys F38252826ACD957EF380D39F2F7956BC5DA04B5D
gpg: directory `/home/tom/.gnupg' created
gpg: new configuration file `/home/tom/.gnupg/gpg.conf' created
gpg: WARNING: options in `/home/tom/.gnupg/gpg.conf' are not yet active during this run
gpg: keyring `/home/tom/.gnupg/secring.gpg' created
gpg: keyring `/home/tom/.gnupg/pubring.gpg' created
gpg: requesting key 5DA04B5D from hkp server pool.sks-keyservers.net
gpg: /home/tom/.gnupg/trustdb.gpg: trustdb created
gpg: key 5DA04B5D: public key "Stanislav Malyshev (PHP key) <smalyshev@gmail.com>" imported
gpg: no ultimately trusted keys found
gpg: Total number processed: 1
gpg:               imported: 1

**Comment by md5** [comment]
I don't recall where this came up before, but it looks like you're running into a networking restriction preventing you from hitting the HKP port 11371. Since some keyservers in the pool support falling back to port 80, your connection will work intermittently. IIRC, someone was able to use 'p80.pool.sks-keyservers.net:80' as the `--keyserver` to get it working. That uses a pool of keyservers that all respond on port 80.

**Comment by md5** [comment]
Here it was: https://github.com/docker-library/mysql/issues/49

**Comment by thomascooper** [comment]
Ok let me try, I actually used the p80 keyserver in a few unsuccessful attempts, but I did not know the hpk protocol needed to be included.

**Comment by tianon** [comment]
Any particular reason you are building from source instead of using the
existing images?

**Comment by thomascooper** [comment]
I have been trying to get PDO to work with the image and am having problems.

**Comment by md5** [comment]
What sort of problems? Did you try something like this?

```Dockerfile
FROM php:5.4-apache
RUN docker-php-ext-install pdo_mysql
```

If you're using a PDO module other than `pdo_mysql`, you may also need to `apt-get install` some dependencies before running `docker-php-ext-install`.

**Comment by thomascooper** [comment]
I cannot get it to load here is my Dockerfile
FROM php:5.4-apache

COPY apache2.conf /etc/apache2/apache2.conf
COPY php.ini /usr/local/etc/php/php.ini

RUN a2enmod rewrite

RUN apt-get update && apt-get install -y \
        libmcrypt-dev \
        libxml2-dev
    && docker-php-ext-install mcrypt \
    && docker-php-ext-install mbstring \
    && docker-php-ext-install soap \
    && docker-php-ext-install bcmath \
    && docker-php-ext-install pdo pdo_mysql

The results of "php -i | grep PDO" are:
PDO
PDO support => enabled
PDO drivers => sqlite
PDO Driver for SQLite 3.x => enabled

The mysql driver is not listed, and my application (which is Kohana using PDO) will not connect with the error:  Database_Exception [ 0 ]: could not find driver

**Comment by thomascooper** [comment]
If I manually log into the box with exec bash and docker-php-ext-install pdo_mysql, it tells me I need a c compiler (so I install build-essential) and then run the install again, it POOF like magic it works. But if I add build-essential to the docker file, after a destroy and up it still not loaded, so I exec bash again, and this time just run the install directly (since build-essential is already installed), the install goes off with no errors and the extension is listed in php -i again.

So basically I am missing something, something is preventing pdo_mysql from being installed by the dockerfile, and im not sure where to look next. So I assumed i would try from source with direct configures.

**Comment by md5** [comment]
Regarding the C compiler error, I think you've hit a bug that I introduced in #71. I've opened #73 as a proposed fix.

I've also confirmed that `pdo_mysql` can be installed against both `php:5.4-cli` and `php:5.4-apache` after the fix and opened a PR to start adding tests for `docker-php-ext-install` in docker-library/official-images#533.

Sorry about that! :disappointed:

**Comment by thomascooper** [comment]
Is there a pre-build image with your fix to #73 applied that i can test with until your change is merged into master?

**Comment by thomascooper** [comment]
Back to the initial issue I had, in trying to build the image locally from your updated commit, I received the same GPG key, as of now updating the server to "hkp://p80.pool.sks-keyservers.net:80" worked and is allowing the image to build, i'll let you know the progress

**Comment by yosifkit** [comment]
Awesome. https://github.com/docker-library/official-images/pull/534 to track once it gets pushed to the docker hub.

**Comment by thomascooper** [comment]
I saw the commit was merged into master, does that mean pulling php:5.4-apache will have the changes, or do i need to wait for the actual images to be updated?

In the mean time I would like to solve this issue I am having still, here are my steps:
* I cloned docker-php from md5 with the commit
* I removed lines 28,45,46 (i am having a problem with gpg even using p80, it build once then blocked me)
* Built thomascooper/php54, pushed to registry
* Created a Dockerfile:
```
FROM thomascooper/php54

COPY apache2.conf /etc/apache2/apache2.conf
COPY php.ini /usr/local/etc/php/php.ini

RUN a2enmod rewrite

RUN apt-get update && apt-get install -y \
        build-essential \
        libmcrypt-dev \
        libxml2-dev \
        php5-mysql
    && docker-php-ext-install mcrypt \
    && docker-php-ext-install mbstring \
    && docker-php-ext-install soap \
    && docker-php-ext-install bcmath \
    && docker-php-ext-install pdo pdo_mysql

WORKDIR /var/www/html/www
CMD ["apache2-foreground"]
```
* Built and ran this Dockerfile via Vagrant, instance came up
* ran docker exec -i-t [name] bash 
```
root@1949a9a88847:/var/www/html/www# php -i | grep PDO
PDO
PDO support => enabled
PDO drivers => sqlite
PDO Driver for SQLite 3.x => enabled
```
* manually ran docker-php-ext-install pdo_mysql
```
Build complete.
Don't forget to run 'make test'.

Installing shared extensions:     /usr/local/lib/php/extensions/no-debug-non-zts-20100525/
grep: /usr/local/etc/php/conf.d/docker-php-ext-pdo_mysql.ini: No such file or directory
```
* checked for PDO again
```
root@1949a9a88847:/var/www/html/www# php -i | grep PDO
PDO
PDO support => enabled
PDO drivers => sqlite, mysql
PDO Driver for MySQL => enabled
PDO Driver for SQLite 3.x => enabled
```
* Pulled an apache served phpinfo()
![no_pdo_mysql](https://cloud.githubusercontent.com/assets/557533/6538114/d1eb8dde-c42b-11e4-9185-1bf63eb356e0.png)

From this it looks like adding back libc-dev allows docker-php-ext-install to run, but it is still not executing from the Dockerfile. 

This leads to the additional problem for me, if I install it directly to the running container, I am not exactly sure the correct way to reload apache to see if that instance at least works

**Comment by thomascooper** [comment]
As a side note, every extension listed in my Dockerfile is missing from 'php -i', not just pdo/pdo_mysql

**Comment by md5** [comment]
I just looked at Docker Hub and saw that it was pushed a few minutes ago (cf. https://registry.hub.docker.com/repos/stackbrew/).

After pulling the new image, I just did this test:

```Dockerfile
FROM php:5.4-apache
RUN docker-php-ext-install pdo_mysql
CMD ["php", "-i"]
```

Then I ran this:

    docker build -t php:test_pdo_mysql .
    docker run --rm php:test_pdo_mysql | grep -i pdo

And got this:

```
Additional .ini files parsed => /usr/local/etc/php/conf.d/ext-pdo_mysql.ini
API Extensions => pdo_mysql
PDO
PDO support => enabled
PDO drivers => sqlite, mysql
pdo_mysql
PDO Driver for MySQL => enabled
pdo_mysql.default_socket => /tmp/mysql.sock => /tmp/mysql.sock
pdo_sqlite
PDO Driver for SQLite 3.x => enabled
```

**Comment by md5** [comment]
As for picking up change to your Apache after "doing things" in the container, you should just be able to do a `docker restart` on the container (though beware you don't corrupt the configuration in the process, lest you have an unstartable container).

**Comment by thomascooper** [comment]
md5: thanks for the help, I was in the process of updating that comment as you replied (a halt/up worked for me)

I will test as soon as I get home, as long as the image works, I really am not worried about why building from source will not work, I am simple

**Comment by yosifkit** [comment]
I see a backslash missing on `libxml2-dev` line; it seems none of your `docker-php-ext-install` should be working.  Too bad docker backpedaled on invalid lines in a Dockerfile...  (I totally have never done this before).   And yes the libc fix is pushed to the Docker Hub for most of the php versions.

**Comment by md5** [comment]
:eyes:

**Comment by thomascooper** [comment]
Yes yes, lets file my idiocy as a bug, not user error please!

After fixing the backslash fixed the not installing problem, but atleast the first bug was real! And this issue can be closed now, thanks for all the help

**Comment by tianon** [comment]
:+1:
